### PR TITLE
merge dev into main

### PR DESCRIPTION
**Pull request by fiftyeightandeight**

### CODE REVIEW

**Comment by JBotwina** [comment]
I think that token-usda will fail on deployment in Alexnet because the addresses aren't seeded in the config file. 
We can make a change by restarting the Alexnet (losing everything so far I think) and changing the config file to include these addresses.

```
(begin
  (try! (ft-mint? usda u1000000000000000 tx-sender))
  (try! (ft-mint? usda u10000000000000 'ST1HTBVD3JG9C05J7HBJTHGR0GGW7KXW28M5JS8QE.alex-reserve-pool))
  (try! (ft-mint? usda u200000000000 'ST1J4G6RR643BCG8G8SR6M2D9Z9KXT2NJDRK3FBTK)) ;;wallet_1
)
```

**Review by JBotwina — APPROVED** [review]
Refer to my earlier comment about token-usda

File: clarity/tests/history.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#V2
```
Comment (hyossid) [inline clarity/tests/history.txt:1]: We should put this file in gitignore

**Comment by hyossid** [comment]
- Checking on weighted equation for price to be always smaller than spot price
- **margin trading using flash loan**
- multisig modification <--- this will conflict with a branch which James and I are working, which we need to modify while merging
- balance-aytoken == balance-token will pass get-price for test cases.

**Comment by hyossid** [comment]
> I think that token-usda will fail on deployment in Alexnet because the addresses aren't seeded in the config file.
> We can make a change by restarting the Alexnet (losing everything so far I think) and changing the config file to include these addresses.
> 
> ```
> (begin
>   (try! (ft-mint? usda u1000000000000000 tx-sender))
>   (try! (ft-mint? usda u10000000000000 'ST1HTBVD3JG9C05J7HBJTHGR0GGW7KXW28M5JS8QE.alex-reserve-pool))
>   (try! (ft-mint? usda u200000000000 'ST1J4G6RR643BCG8G8SR6M2D9Z9KXT2NJDRK3FBTK)) ;;wallet_1
> )
> ```


and the hardcoded address here is not anymore required. Let's remove this hardcoded address as we face it while developing. Only tx-sender may remain for before mainnet launch

**Comment by fiftyeightandeight** [comment]
> > I think that token-usda will fail on deployment in Alexnet because the addresses aren't seeded in the config file.
> > We can make a change by restarting the Alexnet (losing everything so far I think) and changing the config file to include these addresses.
> > ```
> > (begin
> >   (try! (ft-mint? usda u1000000000000000 tx-sender))
> >   (try! (ft-mint? usda u10000000000000 'ST1HTBVD3JG9C05J7HBJTHGR0GGW7KXW28M5JS8QE.alex-reserve-pool))
> >   (try! (ft-mint? usda u200000000000 'ST1J4G6RR643BCG8G8SR6M2D9Z9KXT2NJDRK3FBTK)) ;;wallet_1
> > )
> > ```
> 
> @JBotwina the hardcoded address here is not anymore required. Let's remove this hardcoded address as we face it while developing. Only tx-sender may remain for before mainnet launch

Agreed, but we need them for testing (I use different wallets for different test cases), so let's keep them for now.

File: clarity/contracts/alex-vault.clar [hunk 2]
```diff
@@ -48,13 +43,13 @@
...
-    (unwrap! (contract-call? token transfer amount (as-contract tx-sender) (contract-of flash-loan-user) none) transfer-failed-err)
+    (unwrap! (contract-call? token transfer amount (as-contract tx-sender) tx-sender none) loan-transfer-failed-err)
 
     ;; flash-loan-user executes with loan received
     (unwrap! (contract-call? flash-loan-user execute) user-execute-err)
 
     ;; return the loan + fee
-    (unwrap! (contract-call? token transfer amount-with-fee (contract-of flash-loan-user) (as-contract tx-sender) none) transfer-failed-err) 
+    (unwrap! (contract-call? token transfer amount-with-fee tx-sender (as-contract tx-sender) none) post-loan-transfer-failed-err) 
     (ok true)
```
Comment (hyossid) [inline clarity/contracts/alex-vault.clar:53]: Thinking of it, I think it will be better to return the transferred amount as a tuple as an return value

**Review by hyossid — APPROVED** [review]
I ran the test cases of margin trading in console and values stored seems good. But later on at dev, we can change all (ok true) to some appreciative values

**Comment by hyossid** [comment]
> > > e by restarting the Alexnet (losing everything so far I think) and changing the config file to include these addresses.
> > > ```
> > > (begin
> > >   (try! (ft-mint? usda u1000000000000000 tx-sender))
> > >   (try! (ft-mint? usda u10000000000000 'ST1HTBVD3JG9C05J7HBJTHGR0GGW7KXW28M5JS8QE.alex-reserve-pool))
> > >   (try! (ft-mint? usda u200000000000 'ST1J4G6RR643BCG8G8SR6M2D9Z9KXT2NJDRK3FBTK)) ;;wallet_1
> > > )
> > > ```
> > 
> > 
> > @JBotwina the hardcoded address here is not anymore required. Let's remove this hardcoded address as we face it while developing. Only tx-sender may remain for before mainnet launch
> 
> Agreed, but we need them for testing (I use different wallets for different test cases), so let's keep them for now.

@fiftyeightandeight  Sure, for next test cases, lets make the tx-sender (deployer) distribute tokens (using transfer function in sip10)to other wallets as an initialization of test cases

**Comment by fiftyeightandeight** [comment]
> > > > e by restarting the Alexnet (losing everything so far I think) and changing the config file to include these addresses.
> > > > ```
> > > > (begin
> > > >   (try! (ft-mint? usda u1000000000000000 tx-sender))
> > > >   (try! (ft-mint? usda u10000000000000 'ST1HTBVD3JG9C05J7HBJTHGR0GGW7KXW28M5JS8QE.alex-reserve-pool))
> > > >   (try! (ft-mint? usda u200000000000 'ST1J4G6RR643BCG8G8SR6M2D9Z9KXT2NJDRK3FBTK)) ;;wallet_1
> > > > )
> > > > ```
> > > 
> > > 
> > > @JBotwina the hardcoded address here is not anymore required. Let's remove this hardcoded address as we face it while developing. Only tx-sender may remain for before mainnet launch
> > 
> > 
> > Agreed, but we need them for testing (I use different wallets for different test cases), so let's keep them for now.
> 
> @fiftyeightandeight Sure, for next test cases, lets make the tx-sender (deployer) distribute tokens (using transfer function in sip10)to other wallets as an initialization of test cases

I like the idea - sounds good.

File: clarity/contracts/alex-vault.clar (hunk 2 above)
Comment (fiftyeightandeight) [reply clarity/contracts/alex-vault.clar]: vault now returns uint (= amount-with-fee).

**Pull request merged**